jctools-core/src/main/java/org/jctools/queues/MpscArrayQueue.java

```diff
@@ -331,32 +337,78 @@ public E poll()
         // Copy field to avoid re-reading after volatile load
         final E[] buffer = this.buffer;
 
-        // If we can't see the next available element we can't poll
+        long seenPIndex = seenProducerIndex;
         E e = lvElement(buffer, offset); // LoadLoad
-        if (null == e)
+        if (e != null)
         {
-            /*
-             * NOTE: Queue may not actually be empty in the case of a producer (P1) being interrupted after
-             * winning the CAS on offer but before storing the element in the queue. Other producers may go on
-             * to fill up the queue after this element.
-             */
-            if (cIndex != lvProducerIndex())
+            incrementConsumerIndex(buffer, offset, cIndex, seenPIndex);
+            return e;
+        }
+        if (seenPIndex <= cIndex && (seenPIndex = lvProducerIndex()) == cIndex)
```

Comment:
This reverts the streaming behavior of Fast Flow like algorithm to the cached variant of Lamport's algorithm (introduced by @mjpt777 ), it's a shame to lose out on that one.

Reply:
@nitsanw I don't follow, hoping you misread the logic... the only thing this changes (I think) is to possibly eliminate some (volatile) reads of the producer index by the consumer, via use of a consumer-local cached value. This simpler optimization could actually be done independently of the other changes here, though it becomes more significant when combined with the look-ahead scanning.

I.e. just changing from:
```java
if (cIndex == lvProducerIndex()) { return null; }
```
to
```java
if (cIndex >= cachedProducerIndex && cIndex == lvProducerIndex()) { return null; }
```